From d08083d057d6cc7136128cad6ecefba43b886c4c Mon Sep 17 00:00:00 2001
From: Vishal Pandey <vpandey@redhat.com>
Date: Thu, 9 May 2019 14:37:22 +0530
Subject: [PATCH 138/141] glusterd: Add gluster volume stop operation to
 glusterd_validate_quorum()

ISSUE: gluster volume stop succeeds even if quorum is not met.

Fix: Add GD_OP_STOP_VOLUME to gluster_validate_quorum in
glusterd_mgmt_v3_pre_validate ().

Since the volume stop command has been ported from synctask to mgmt_v3,
the quorum check was missed out.

>upstream patch : https://review.gluster.org/#/c/glusterfs/+/22692/

>Change-Id: I7a634ad89ec2e286ea262d7952061efad5360042
>fixes: bz#1690753
>Signed-off-by: Vishal Pandey <vpandey@redhat.com>

BUG: 1706893
Change-Id: I7a634ad89ec2e286ea262d7952061efad5360042
Signed-off-by: Vishal Pandey <vpandey@redhat.com>
Reviewed-on: https://code.engineering.redhat.com/gerrit/169949
Tested-by: RHGS Build Bot <nigelb@redhat.com>
Reviewed-by: Sunil Kumar Heggodu Gopala Acharya <sheggodu@redhat.com>
---
 tests/bugs/glusterd/quorum-validation.t   | 4 +++-
 xlators/mgmt/glusterd/src/glusterd-mgmt.c | 2 +-
 2 files changed, 4 insertions(+), 2 deletions(-)

diff --git a/tests/bugs/glusterd/quorum-validation.t b/tests/bugs/glusterd/quorum-validation.t
index 05aef4e..ff46729 100644
--- a/tests/bugs/glusterd/quorum-validation.t
+++ b/tests/bugs/glusterd/quorum-validation.t
@@ -34,9 +34,11 @@ TEST ! $CLI_1 volume add-brick $V0 $H1:$B1/${V0}2
 TEST ! $CLI_1 volume remove-brick $V0 $H1:$B1/${V0}0 start
 TEST ! $CLI_1 volume set $V0 barrier enable
 
-# Now execute a command which goes through op state machine and it should fail
 TEST ! $CLI_1 volume profile $V0 start
 
+#bug-1690753 - Volume stop when quorum not met is successful
+TEST ! $CLI_1 volume stop $V0
+
 #Bring back the 2nd glusterd
 TEST $glusterd_2
 
diff --git a/xlators/mgmt/glusterd/src/glusterd-mgmt.c b/xlators/mgmt/glusterd/src/glusterd-mgmt.c
index 61ad66e..ec78913 100644
--- a/xlators/mgmt/glusterd/src/glusterd-mgmt.c
+++ b/xlators/mgmt/glusterd/src/glusterd-mgmt.c
@@ -1059,7 +1059,7 @@ glusterd_mgmt_v3_pre_validate(glusterd_op_t op, dict_t *req_dict,
         goto out;
     }
 
-    if (op == GD_OP_PROFILE_VOLUME) {
+    if (op == GD_OP_PROFILE_VOLUME || op == GD_OP_STOP_VOLUME) {
         ret = glusterd_validate_quorum(this, op, req_dict, op_errstr);
         if (ret) {
             gf_msg(this->name, GF_LOG_ERROR, 0, GD_MSG_SERVER_QUORUM_NOT_MET,
-- 
1.8.3.1

